certbot_route53/authenticator.py

```diff
@@ -14,112 +16,128 @@
-            ):
-                break
-            logger.info("Waiting for DNS propagation...")
-            time.sleep(1)
-        else:
-            logger.error("Unable to verify domain " + achall.domain)
-            return None
-
-        return response
-
-    def cleanup(self, achalls):
-        # pylint: disable=missing-docstring
-        r53 = boto3.client('route53')
-        for achall in achalls:
-            try:
-                zone = self._find_zone(r53, achall.domain)
-            except ValueError:
-                logger.warn("Unable to find zone for " + achall.domain + ". Skipping cleanup.")
-                continue
-
-            _, validation = achall.response_and_validation()
-            self._excute_r53_action(r53, achall, zone, validation, 'DELETE')
-        return None
-
-
-    def _excute_r53_action(self, r53, achall, zone, validation, action, wait_for_change=False):
-            response = r53.change_resource_record_sets(
-                HostedZoneId=zone["Id"],
-                ChangeBatch={
-                    'Comment': 'Let\'s Encrypt ' + action,
-                    'Changes': [
-                        {
-                            'Action': action,
-                            'ResourceRecordSet': {
-                                'Name': achall.validation_domain_name(achall.domain),
-                                'Type': 'TXT',
-                                'TTL': TTL,
-                                'ResourceRecords': [
-                                    {
-                                        'Value': '"' + validation + '"',
-                                    },
-                                ],
-                            },
-                        },
-                    ],
-                },
+            change_ids = [self._create_single(achall) for achall in achalls]
+            for change_id in change_ids:
+                self._wait_for_change(change_id)
+            return [achall.response(achall.account_key) for achall in achalls]
+        except NoCredentialsError:
+            raise Exception("No AWS Route53 credentials found. " + INSTRUCTIONS)
+
+    def cleanup(self, achalls):  # pylint: disable=missing-docstring
+        for name, value in self.txt_records:
+            self._delete_txt_record(name, value)
+
+    def _create_single(self, achall):
+        """Create a TXT record, return a change_id"""
+        name, value = (achall.validation_domain_name(achall.domain),
+            achall.validation(achall.account_key))
+        change_id = self._create_txt_record(name, value)
+        self.txt_records.append((name, value))
+        return change_id
+
+    def _find_zone_id_for_domain(self, domain):
+        paginator = self.route53_client.get_paginator("list_hosted_zones")
+        zones = []
+        for page in paginator.paginate():
+            for zone in page["HostedZones"]:
+                if (
+                    domain.endswith(zone["Name"]) or
```

Comment:
Good catch. I pulled this in from letsencrypt-aws and didn't properly reconcile it with your search version. But you're right that this has a bug. I'll implement something like what you suggest here.